### ISSUE TITLE
.cfignore /xxxx/ lines are ignored in 6.34.0 and in the edge binary to fix #1300

### ISSUE THREAD

**Issue by rlk833** [cloudfoundry/cli#1305]
## Command  

cf push dev-resource-catalog-blue -f manifest.yml

## What occurred

The `/xxxx/` type .cfignore were not processed. The folders that should of been ignored weren't ignored and were pushed. 

## What you expected to occur

That the xxxx folders found in the current directory that cf push was called from should of been ignored. They weren't properly ignored, as per cf gitignore documentation. Worked in previous versions. 

But , /xxxx for folders does work.

## CLI Version  

cf version 6.34.0+b1f628e0f.2018-01-11

(I was testing the patch fix from #1300 to see if it fixed this .cfignore problem too)

## CC API Endpoint Version  

2.92.0

## CF Trace

n/a

## Platform & Shell Details  

Ubuntu image in travis. I don't know the version

**Comment by XenoPhex** [comment]
@rlk833 We're looking into it now, but are you seeing the entire folder get uploaded or is it uploading just an empty folder?

**Comment by rlk833** [comment]
It was uploading contents too.

**Comment by XenoPhex** [comment]
Can you give us an example layout with a `.cfignore`? We're trying to reproduce it and what we're seeing (with the edge binary) is that it's uploading an empty folder when we have a `/` at the end of a folder name.

**Comment by rlk833** [comment]
/xyz/  is what I was using. Note that /xyz worked. It has to have a leading '/' and a trailing slash to fail.

I hadn't tried just xyz/  because it was necessary that it is from the root of the push tree and not some lower folder that happened to be called xyz.

**Comment by XenoPhex** [comment]
Using the latest edge build we've tried the following:

| starting slash | trailing slash | result |
| ------------------ | ---------------- | -------- |
| :x:  | :x: | will ignore that file/directory |
| :heavy_check_mark:  | :x: | will ignore that file/directory |
| :heavy_check_mark:  | :heavy_check_mark: | (only works for directories) ignores the contents but selects the folder itself |
| :x:  | :heavy_check_mark: | (only works for directories) ignores the contents but selects the folder itself |

Additionally, can you try `cf push`ing with `CF_LOG_LEVEL=debug` before the push itself? It'll output a LOT of information, but you should see a section that looks like the following:
```
INFO[0000] creating archive                             
INFO[0000] zipping source files from directory           sourceDir=/home/pivotal/go/src/github.com/cloudfoundry/cf-acceptance-tests/assets/dora
DEBU[0000] zipping file                                  fullPath=/home/pivotal/go/src/github.com/cloudfoundry/cf-acceptance-tests/assets/dora/.rspec
DEBU[0000] resource file info                            file-mode=-rw-r--r--
DEBU[0000] setting mode for file                         destPath=.rspec mode=-rw-r--r-- srcPath=/home/pivotal/go/src/github.com/cloudfoundry/cf-acceptance-tests/assets/dora/.rspec
DEBU[0000] zipping file                                  fullPath=/home/pivotal/go/src/github.com/cloudfoundry/cf-acceptance-tests/assets/dora/Gemfile
DEBU[0000] resource file info                            file-mode=-rw-r--r--
...
```
If the edge binary is uploading only the folder, you'll see an entry for the folder and nothing underneath it. If it's uploading it's contents, you should see `cf` zipping the contents.

Sorry if we're being repetitive, we're just trying to reproduce what you're seeing and we're having problems doing so.

**Comment by dkoper** [comment]
Hi @rlk833 

I wonder if what you are pointing out is an issue with our documentation, in particular the following:
http://docs.cloudfoundry.org/devguide/deploy-apps/prepare-to-deploy.html

> For example, these lines in a .cfignore file exclude the “tmp” and “log” directories.
```
tmp/
log/
```

It looks like we forgot to update this: now that we follow `.gitignore` rules, this example would exclude the files **under** those folder names, but would still include the (empty) folders themselves.
I have submitted a PR to the docs team to address that.
https://github.com/cloudfoundry/docs-dev-guide/pull/223

**Comment by rlk833** [comment]
Now getting these results, slightly different, but more sane. :-)
Basically if there is a trailing slash (doesn't matter whether leading slash or not) you get folder but no contents.

```cf version 6.34.0+65c3b26dc.2018-01-11

All of the  folders in this example had content before pushing

1)  /tests/  - After pushing only name was pushed,  not contents

DEBU[0030] zipping file                                  fullPath=/home/travis/gopath/src/github.ibm.com/Bluemix/platform-resource-catalog/tests
DEBU[0030] resource file info                            file-mode=drwxrwxr-x
DEBU[0030] setting mode for file                         destPath=tests/ mode=drwxr-xr-x srcPath=/home/travis/gopath/src/github.ibm.com/Bluemix/platform-resource-catalog/tests

2) dbdocker/ - After pushing only name was pushed, not contents

DEBU[0026] zipping file                                  fullPath=/home/travis/gopath/src/github.ibm.com/Bluemix/platform-resource-catalog/dbdocker
DEBU[0026] resource file info                            file-mode=drwxrwxr-x
DEBU[0026] setting mode for file                         destPath=dbdocker/ mode=drwxr-xr-x srcPath=/home/travis/gopath/src/github.ibm.com/Bluemix/platform-resource-catalog/dbdocker

3) docs - Neither name nor contents pushed

--- No DEBU messages at all for it

4) /manifest - Neither name nor contents pushed

--- No DEBU messages at all for it```

**Comment by jbpivotal** [comment]
@rlk833 It sounds like the clarification helped. I'm planning to close this issue as it sounds like the documentation change is all we need. Let me know if there's anything else.

**Comment by rlk833** [comment]
No problem. Just a change in definition. We have a lot /ccc/ and we just need to be aware this now eliminates the contents but not the folder. That's not too bad, an empty folder doesn't cause problems like the contents would.

Just waiting on this fix being in a release so I can go back to getting the latest stable instead of the edge builds.

**Comment by jbpivotal** [comment]
@rlk833 We've included the fix to 1300 in 6.34.1. I'll close this issue.
